### ISSUE TITLE
[Epic] Separate source connector out of source description to allow re-use of connection details between sources

### ISSUE THREAD

**Issue by nharring-adjacent** [MaterializeInc/materialize#11504]
### Work items

* [x] #12982

* [x] #12983

* [x] Rename `CREATE CONNECTOR` to `CREATE CONNECTION` (#13060)

* [x] #13017
    Addressed by WIP PR #13103.

* [x] Share connector struct definitions, parsing, and planning between `CREATE SOURCE` and `CREATE CONNECTION`. The `KafkaSourceConnection` struct should embed a `KafkaConnection` struct, for example.
    Addressed by WIP PR #13103.

* [x] https://github.com/MaterializeInc/materialize/issues/13012
   Awaiting Nikhil's time.

* [x] Redesign `CREATE CONNECTION` options to use @sploiselle's new "keyword options."
    Sean is actively working on this: #13062.

* [x] Add connectors for remaining source types
    Blocked on the above.

* [ ] Deprecate the existing inline connection experience. (#13374)

* [ ] #13592

* [x] #13675

### Feature request

Motivation sourced from the design doc draft:

Many users of Materialize create families of sources and sinks that share many
configuration parameters. The current design of `CREATE SOURCE` and `CREATE
SINK` make this a verbose affair. The problem is particularly acute with
Avro-formatted Kafka sources that configure authentication:

```sql
CREATE SOURCE kafka1
FROM KAFKA BROKER 'kafka:9092' TOPIC 'top1' WITH (
    security_protocol = 'SASL_SSL',
    sasl_mechanisms = 'PLAIN',
    sasl_username = 'username',
    sasl_password = 'password',
)
FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
    username = 'username',
    password = 'password'
);
CREATE SOURCE kafka2
FROM KAFKA BROKER 'kafka:9092' TOPIC 'top2' WITH (
    security_protocol = 'SASL_SSL',
    sasl_mechanisms = 'PLAIN',
    sasl_username = 'username',
    sasl_password = 'password'
)
FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY 'https://schema-registry' WITH (
    username = 'username',
    password = 'password'
);
```

These two source definition differ only in their topic specification, but
must duplicate eight other configuration parameters.

There is a lot of opportunity for improving the user experience, starting with streamlining source creation but also future iterations could enable updating credentials without recreating a whole source, and also create opportunities for improved efficiency at the source implementation layer.

**Comment by nharring-adjacent** [comment]
I am thinking that to get `CREATE CONNECTOR` for Kafka Broker and CSR through `experimental` status and into stable shape it breaks down into the following items and their time estimates

- [ ] Add `CREATE CONNECTOR` statement using existing syntax for Broker and CSR options behind experimental flag -- 3 days
- [ ] Add optional field for ID relation to KafkaSourceConnector in the kafka variant of ExternalSourceConnector -- .5 days
- [ ] Update sqlparser to accept `CREATE SOURCE .. FROM CONNECTOR` syntax -- 2 days
- [ ] Add logic in `plan_create_source` for Kafka sources to create implicit connector and add source to catalogue using that link -- 5 days
- [ ] Update syntax for `CREATE CONNECTOR` for Kafka Broker and CSR to something like what the [design](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20220404_connectors.md#kafka-connector) specifies -- 5-7 days
  - [ ] Add catalog migration from old to new connector syntax -- 1 day
- [ ] Update syntax for `CREATE SOURCE .. KAFKA BROKER ..` to conform to new connector syntax -- 3-4 days

**Comment by nmeagan11** [comment]
Per [internal Slack discussion](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1653664730737289), we will change the syntax from `CREATE CONNECTOR` to `CREATE CONNECTION`.

**Comment by nmeagan11** [comment]
Anything left to do here? Can we close this? (cc @sploiselle @benesch)

**Comment by sploiselle** [comment]
@nmeagan11 The remaining work is

> Deprecate the existing inline connection experience.

This is largely tracked in #13374, so this could be closed and we could just track that as the remaining work item. I know it's also part of the ongoing `SINK` work, too.

**Comment by sploiselle** [comment]
Closing this in favor of #13374, which has slightly less history to wade through.
